### ISSUE TITLE
Move to modulegraph2

### ISSUE THREAD

**Issue by htgoebel** [pyinstaller/pyinstaller#4535]
`modulegraph2` is a complete rewrite of `modulegraph` (by the same author), using lessons learned in that project but with a complete new Python 3 code base and full test coverage.

`modulegraph` will no longer be maintained.

We need to identify our changes to `modulegraph`, identify which are relevant for `modulegraph2` and port them over.

I marked this as "good first issue", since the changes are focused on a single part of the code, which has good code-coverage. I have not yet had a look at the code of `modulegraph2`, though, so I can not tell whether we need to change a lot.

When implementing this, you might need many cycles to get it right
- please work on a separate branch, e.g. `modulegraph2`
- run the test-suite locally on your system prior to pushing it anwhere
- set up your personal travis-ci and appveyor and push there first
- open the pull-request only when everything is fine and all tests pass in your travis-ci and appveyor.

This helps keeping our CI "credits" for your main development up to speed (it's slow anyway) until the code is ready. Thanks!

**Comment by htgoebel** [comment]
Modulegraph2 required Python 3.6, so we can only switch after dropping support for Python 3.5.
Homepage: https://modulegraph2.readthedocs.io/

**Comment by Legorooj** [comment]
@htgoebel I'd like to head this development branch if you don't mind. Once I've finished the hooks-contrib move that won't need too much attention.

**Comment by nehaljwani** [comment]
I have observed that py38 support has been merged into develop and I don't see any py38 specific issues in the milestone for 4.1.  Would it be possible to cut a release to officially support py38 and put the move to modulegraph2 in the milestone of the next release?

**Comment by Legorooj** [comment]
@nehaljwani that's the plan, the milestones just haven't been updated yet. Waiting on @htgoebel to release.

**Comment by htgoebel** [comment]
@Legorooj We could start with this after PyInstaller 4.2 release, with also dropping support for Py 3.5 after PyI 4.2 release - see #5430
I remember you started working in modulegraph2 but stepped over some deficits or issues?!

**Comment by Legorooj** [comment]
Yes, I started work and found quite a few bugs (most of which have been fixed now), plus modulegraph2 doesn't support python 3.5. I'm going to look into some solutions such as submodules and subclassing modulegraph instead of including our own copy that needs to be updated. I'll be starting work within the next week or two, and plan to have this ready to go for 4.3. 

A lot of the edge cases that we bolted handlers for onto modulegraph1 are now supported by default in modulegraph2, so hopefully it should require minimal changes to get it working.

**Comment by aqeelat** [comment]
Hi,
I came across this issue from https://goodfirstissue.dev/

I see that the last activity on this issue was a while ago. Is @Legorooj still working on it? If not, can I pick it up?

**Comment by bwoodsend** [comment]
Contrary to the label, this really isn't a good first issue. And yes, Legorooj is still working on it.
